jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -264,14 +285,23 @@ public static void copy(Resource resource, Content.Sink sink, ByteBufferPool buf
         // Directly write the byte array if the resource is a MemoryResource.
         if (resource instanceof MemoryResource memoryResource)
         {
-            byte[] bytes = memoryResource.getBytes();
-            sink.write(true, ByteBuffer.wrap(bytes), callback);
+            sink.write(true, ByteBuffer.wrap(memoryResource.getBytes()), callback);
             return;
         }
 
         // Fallback to Content.Source.
-        Content.Source source = asContentSource(resource, bufferPool, bufferSize, direct);
-        Content.copy(source, sink, callback);
+        try
+        {
+            InputStream inputStream = resource.newInputStream();
+            if (inputStream == null)
+                throw new IllegalArgumentException("Resource does not support InputStream: " + resource);
+            Content.Source source = Content.Source.from(new ByteBufferPool.Sized(bufferPool, direct, bufferSize), inputStream, 0, -1);
+            Content.copy(source, sink, callback);
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeIOException(e);
```

Comment:
Cannot throw, there is a callback parameter that must be failed instead.

Reply:
I agree the signature is a little strange having a throw and a callback, but I think it is clear that the throw is only for illegal arguments.   So I agree the callback should be failed here.

Reply:
You are right, the two `copy()` variants take a callback so they should never throw. Let me fix that.